docs(S3): perf optimization guidelines MTA-3845 (#2448)
scaleway/docs-content: storage/object/reference-content/optimize-object-storage-performance.mdx

For example, if the most CPU-intensive operation uses 20% of your CPU, you can execute up to five requests concurrently, to reach 100% of your processing capacity.

<Message type="tip">
  Several tools, such as [htop](https://htop.dev/) allow you to monitor your CPU activity in real time to evaluate the strain of your Object Storage workloads.
</Message>

### Geographic location

The physical distance to the hardware hosting your Object Storage can also have an impact on performance, especially on latency. Make sure to benchmark the different [regions](/storage/object/concepts/##region-and-availability-zone) where Object Storage is available to compare latency on your mission-critical S3 operations.

For instance, media and content distribution are often heavily affected by the physical distance between the host and the client, as objects are usually large in this scenario.

<Message type="tip">

> To optimize latency and throughput on `GET` operations, Scaleway’s [Edge Services](/storage/object/how-to/get-started-edge-services/) allows for closer caching of your objects.

Reviewer: This is currently in Private Beta by invitation only - the link to the doc has only been given to beta testers. I think it would be better to either not mention it, or remove the link to the doc and say something along the lines of "This is currently available to private beta testers only - [get in touch](https://www.scaleway.com/en/betas/#edge-services) if you're interested in joining the beta"

---
meta:
  title: Optimize your Object Storage performance

>   description: This page presents recommandations to optimize your Object Storage performance.

Reviewer 2: ```suggestion
  description: This page presents recommendations to optimize your Object Storage performance.
```

content:
  h1: Optimize your Object Storage performance

>   paragraph: This page presents recommandations to optimize your Object Storage performance.

Reviewer 2: ```suggestion
  paragraph: This page presents recommendations to optimize your Object Storage performance.
```

tags: object storage object-storage performance tuning better enhance throughput latency optimize
dates:
  validation: 2023-11-24

>   posted: 2021-11-24

Reviewer 2: ```suggestion
  posted: 2023-11-24
```

categories:
  - storage
  - object-storage
---

[Scaleway Object Storage](/storage/object/concepts/#object-storage) is a highly resilient and versatile service that guarantees the reliability and accessibility of your data, while being fully [S3-compatible](/storage/object/concepts/#s3) and user-friendly.

Even though it is designed to provide best-in-class latency and throughput, user infrastructure plays a predominant role in achieving optimum efficiency, as many different factors can have an impact on performance, such as your hardware, your software stack, or the way you manage your objects.

> These guidelines aim at providing recommendations and methods to help you get the most out of your Object Storage.

Reviewer 2: ```suggestion
These guidelines aim to provide recommendations and methods to help you get the most out of your Object Storage.
```

Adapt the throughput and latency of your bandwidth according to your most critical use cases:

- Favor throughput over latency for concurrent requests on large multipart requests to maximize efficiency. A smaller number of objects means less per-request connection setups and Secure Socket Layer (SSL) handshakes, which generate latency for each part, but have little-to-no impact on throughput thanks to parallelization.

- Favor latency over throughput for smaller-size objects (less than 1 MB). Avoid per-request connection to remove TCP slow starts and SSL handshakes on each request. A lower latency will also reduce your Time To First Byte (TTFB) for larger objects. The tradeoff is that `ListObject` operations will take longer to complete as they will return a greater number of objects. 

<Message type="tip">
  [Prefixes](/storage/object/concepts/#prefix) and the `MaxKeys` parameter allow you to target smaller object batches.
</Message>

### CPU

The CPU plays a major role in your Object Storage infrastructure, as it will determine your concurrency capacity. A good order of magnitude is 16 cores per socket to handle concurrent requests efficiently.

> Try various operations with a single request at a time and measure your CPU usage. The operations that have the highest CPU usage are likely to throttle your machine and create bottlenecks. These operations can be used to determine the concurrency capacity of your object storage deployment.

Reviewer 2: ```suggestion
Try various operations with a single request at a time and measure your CPU usage. The operations that have the highest CPU usage are likely to throttle your machine and create bottlenecks. These operations can be used to determine the concurrency capacity of your Object Storage deployment.
```

> For example, if the most CPU-intensive operation uses 20 % of your CPU, you can execute up to five requests concurrently, to reach 100 % of your processing capacity.

Reviewer 2: ```suggestion
For example, if the most CPU-intensive operation uses 20% of your CPU, you can execute up to five requests concurrently, to reach 100% of your processing capacity.
```

<Message type="tip">
  Several tools, such as [htop](https://htop.dev/) allow you to monitor your CPU activity in real time to evaluate the strain of your Object Storage workloads.
</Message>

### Geographic location

> The physical distance to the hardware hosting your Object Storage can also have an impact on performance, especially on latency. Make sure to benchmark several regions to compare latency on your mission critical S3 operations.

Reviewer 2: ```suggestion
The physical distance to the hardware hosting your Object Storage can also have an impact on performance, especially on latency. Make sure to benchmark several regions to compare latency on your mission-critical S3 operations.
```

Media and content distribution are often heavily affected by the physical distance between the host and the client, as objects are usually large in this scenario.

<Message type="tip">

> To optimize latency and throughput on `GET` operations, Scaleway’s Edge Services allows for closer caching of your objects.

Reviewer 2: ```suggestion
To optimize latency and throughput on `GET` operations, Scaleway’s [Edge Services](/storage/object/how-to/get-started-edge-services/) allows for closer caching of your objects.
```

</Message>

## Software recommendations

Now that your hardware is properly set, make sure that your object transfers are properly set. The way you manage your object operations also has a major impact on performance, as well as the tools you use.

### Multipart uploads

> The most impactful feature is multipart uploads. Large and small objects have different constraints, and adapting your processes to your “raw material” is quintessential for a seamless object storage deployment. Breaking bigger objects into multiple parts will allow you to leverage the full potential of Scaleway Object storage.

Reviewer 2: ```suggestion
The most impactful feature is multipart uploads. Large and small objects have different constraints, and adapting your processes to your “raw material” is quintessential for a seamless Object Storage deployment. Breaking bigger objects into multiple parts will allow you to leverage the full potential of Scaleway Object Storage.
```

Multipart uploads must respect the following technical specifications:

- 1 to 1000 parts per object
- 5 MB to 5 GB per part (except for the last part)
- Object size cannot exceed 5 TB

<Message type="tip">
  Consider using multipart uploads for objects larger than 50 MB. refer to [Managing multipart uploads](/storage/object/api-cli/multipart-uploads/) for more information.
</Message>

Breaking your objects into very small parts is counter productive, as many operations are executed on each part, thus increasing the latency and reducing the throughput.

Multipart upload allows you to leverage an important concurrency capacity, because:

> - you can **parallelize parts upload** for a better throughput.

Reviewer 2: ```suggestion
- you can **parallelize part uploads** for a better throughput.
```

> smaller objects parts reduce the risk of failed upload due to a network issue, as it allows to retry the upload of the failed part and not the whole object.

Reviewer 2: ```suggestion
smaller object parts reduce the risk of failed uploads due to a network issue, as it allows to retry the upload of the failed part and not the whole object.
```

- you can **pause your upload and resume later**, and upload parts over a desired period of time to reduce stress linked to upload on your infrastructure.

- you can **upload object parts as you are creating them** without knowing the size of the object or the number of parts.

- you can **request different parts simultaneously** using a single cURL request. `GetObject` requests allow you to know the number of parts of an object.

<Message type="tip">
Use the `Range` header in `GET` requests (both for multipart and non-multipart objects) to create concurrent requests for a better throughput.
</Message>

### Tools and features

Try benchmarking several clients to upload objects, such as [MinIO CLI](/storage/object/api-cli/installing-minio-client/), [awscli](/storage/object/api-cli/object-storage-aws-cli/), [Rclone](/storage/object/api-cli/installing-rclone/), or [s5cmd](https://github.com/peak/s5cmdawscli).

Some clients may not be CLI tools, but rather SDKs, which can add slowness in managing Object Storage resources. do not hesitate in testing other tools.

Leverage the different [storage classes](/storage/object/concepts/#storage-class) offered by Scaleway, and benchmark

> Enable [Scalerway Cockpit](/observability/cockpit/quickstart/) to access detailed and in-depth monitoring and logs to identify potential bottlenecks in your Object Storage deployment.

Reviewer 2: ```suggestion
Enable [Scaleway Cockpit](/observability/cockpit/quickstart/) to access detailed and in-depth monitoring and logs to identify potential bottlenecks in your Object Storage deployment.
```

Enable [Scaleway Edge Services](/storage/object/how-to/get-started-edge-services/) to achieve better performance on `GET` 
operations by caching your objects closer to your users. This feature is particularly adapted to CDN applications.

> Throughput is limited at the bucket level to guarantee the security and stability of Scaleway’s Object storage infrastructure. If you notice recurring rate limit errors (error 429) in your [Cockpit](/observability/cockpit/quickstart/) logs, [contact Scaleway’s support](/console/account/how-to/open-a-support-ticket/) to unlock the threshold for your Organization.

Reviewer 2: ```suggestion
Throughput is limited at the bucket level to guarantee the security and stability of Scaleway’s Object Storage infrastructure. If you notice recurring rate limit errors (error 429) in your [Cockpit](/observability/cockpit/quickstart/) logs, [contact Scaleway’s support](/console/account/how-to/open-a-support-ticket/) to unlock the threshold for your Organization.
```

For example, if the most CPU-intensive operation uses 20% of your CPU, you can execute up to five requests concurrently, to reach 100% of your processing capacity.

<Message type="tip">
  Several tools, such as [htop](https://htop.dev/) allow you to monitor your CPU activity in real time to evaluate the strain of your Object Storage workloads.
</Message>

### Geographic location

The physical distance to the hardware hosting your Object Storage can also have an impact on performance, especially on latency. Make sure to benchmark the different [regions](/storage/object/concepts/##region-and-availability-zone) where Object Storage is available to compare latency on your mission-critical S3 operations.

For instance, media and content distribution are often heavily affected by the physical distance between the host and the client, as objects are usually large in this scenario.

<Message type="tip">

> To optimize latency and throughput on `GET` operations, Scaleway’s [Edge Services](/storage/object/how-to/get-started-edge-services/) allows for closer caching of your objects. This feature is currently in private beta, register [here](https://www.scaleway.com/en/betas/#edge-services) if you are interested in joining in.

Reviewer: ```suggestion
To optimize latency and throughput on `GET` operations, Scaleway’s Edge Services allows for closer caching of your objects. This feature is currently in private beta, register [here](https://www.scaleway.com/en/betas/#edge-services) if you are interested in joining in.
```

</Message>

## Software recommendations

Now that your hardware is properly set up, make sure that your object transfers are properly configured. The way you manage your object operations also has a major impact on performance, as well as the tools you use.

### Multipart uploads

The most impactful feature is multipart uploads. Large and small objects have different constraints, and adapting your processes to your “raw material” is quintessential for a seamless Object Storage deployment. Breaking bigger objects, such as uncompressed 4K videos, into multiple parts will allow you to leverage the full potential of Scaleway Object Storage.

Multipart uploads must respect the following technical specifications:

- 1 to 1000 parts per object
- 5 MB to 5 GB per part (except for the last part)
- Object size cannot exceed 5 TB

<Message type="tip">

>   Consider using multipart uploads for objects larger than 50 MB. refer to [Managing multipart uploads](/storage/object/api-cli/multipart-uploads/) for more information.

Reviewer 3: ```suggestion
  Consider using multipart uploads for objects larger than 50 MB. Refer to [our documentation on managing multipart uploads](/storage/object/api-cli/multipart-uploads/) for more information.
```

- you can **upload object parts as you are creating them**, without knowing the size of the object or the number of parts.

- you can **request different parts simultaneously** using a single cURL request. `GetObject` requests allow you to know the number of parts of an object.

<Message type="tip">
Use the `Range` header in `GET` requests (both for multipart and non-multipart objects) to create concurrent requests for a better throughput.
</Message>

### Tools and features

Try benchmarking several clients to upload objects, such as [MinIO CLI](/storage/object/api-cli/installing-minio-client/), [awscli](/storage/object/api-cli/object-storage-aws-cli/), [Rclone](/storage/object/api-cli/installing-rclone/), or [s5cmd](https://github.com/peak/s5cmdawscli).

Leverage the different [storage classes](/storage/object/concepts/#storage-class) offered by Scaleway, and benchmark them according to your needs to find the right configuration for your deployment.

Enable [Scaleway Cockpit](/observability/cockpit/quickstart/) to access detailed and in-depth monitoring and logs to identify potential bottlenecks in your Object Storage deployment.

> Enable [Scaleway Edge Services](/storage/object/how-to/get-started-edge-services/) to achieve better performance on `GET` 

Reviewer: ```suggestion
Enable Scaleway Edge Services to achieve better performance on `GET` 
```
